File: ruby/strain/strain.rb
class Array
  def keep
    return to_enum(:keep) unless block_given?

    result = []
    index = 0
    while index < length
      result << self[index] if yield(self[index])
      index += 1
    end
    result
  end

  def discard(&block)
    return to_enum(:discard) unless block_given?

    self - keep(&block)
  end
end

File: javascript/strain/strain.js
function keep(collection, callback) {
  let item
  const result = []
  for(let i = 0; i < collection.length; i++) {
    item = collection[i]
    if (callback(item)) {
      result.push(collection[i])
    }
  }

  return result
}

function discard(collection, callback) {
  return keep(collection, negate(callback))
}

function negate(callback) {
  return function(item) {
    return !callback(item)
  }
}

module.exports = {
  keep,
  discard
}
